File: agentic-clj/test/agentic/suspend_test.clj
(ns agentic.suspend-test
  "Human-in-the-loop suspend/resume goldens — mirrors the jagentic-core HumanGate tests."
  (:require [clojure.test :refer [deftest is testing]]
            [agentic.core :as core]
            [agentic.event :as ev]
            [agentic.suspend :as sus]))

(def ^:private needs-approval?
  #(contains? (:metadata %) "needs_approval"))

(defn- approval-event []
  (ev/event "c1" "alice" "what is my balance?" {"needs_approval" "true"}))

(deftest approval-submit-suspends
  (testing "an approval-flagged turn suspends instead of completing"
    (let [svc (sus/suspension-service)
          g (sus/human-gate (core/banking-system) svc needs-approval?)
          res (sus/gate-submit g (approval-event) 0)]
      (is (= "awaiting-approval" (:path res)))
      (is (false? (:ok res)))
      (is (true? (sus/suspended? svc "c1")))
      (is (= [] (:tool-calls res))))))

(deftest resume-approved-replays-turn
  (testing "approved resume replays the held turn through the real graph"
    (let [svc (sus/suspension-service)
          g (sus/human-gate (core/banking-system) svc needs-approval?)]
      (sus/gate-submit g (approval-event) 0)
      (let [res (sus/gate-resume g "c1" true 10)]
        (is (= "payments" (:path res)))
        (is (re-find #"1234\.56" (:reply res)))
        (is (false? (sus/suspended? svc "c1")))))))

(deftest resume-denied-reports
  (testing "denied resume reports and clears the suspension"
    (let [svc (sus/suspension-service)
          g (sus/human-gate (core/banking-system) svc needs-approval?)]
      (sus/gate-submit g (approval-event) 0)
      (let [res (sus/gate-resume g "c1" false 10)]
        (is (= "denied" (:path res)))
        (is (false? (:ok res)))
        (is (false? (sus/suspended? svc "c1")))))))

(deftest normal-turn-passes-through
  (testing "a turn without the approval flag passes straight through the gate"
    (let [svc (sus/suspension-service)
          g (sus/human-gate (core/banking-system) svc needs-approval?)
          res (sus/gate-submit g (ev/event "c1" "alice" "what is my balance?") 0)]
      (is (= "payments" (:path res)))
      (is (true? (:ok res)))
      (is (false? (sus/suspended? svc "c1"))))))

(deftest timeout-escalates
  (testing "a suspension past the timeout escalates; before the timeout, nothing"
    (let [svc (sus/suspension-service)
          g (sus/human-gate (core/banking-system) svc needs-approval? 1000)]
      (sus/gate-submit g (approval-event) 0)
      (is (= [] (sus/gate-check-timeouts g 500)))
      (let [out (sus/gate-check-timeouts g 2000)]
        (is (= 1 (count out)))
        (is (= "escalated" (:path (first out))))
        (is (false? (sus/suspended? svc "c1")))))))

(deftest resume-nothing-pending
  (testing "resume with no pending suspension reports nothing pending"
    (let [svc (sus/suspension-service)
          g (sus/human-gate (core/banking-system) svc needs-approval?)
          res (sus/gate-resume g "c1" true 10)]
      (is (= "resume" (:path res)))
      (is (false? (:ok res))))))

File: ports/experimental/go/core/suspend_test.go
package core

import (
	"strings"
	"testing"
)

// approvalNeeds is the gate predicate: an event needs approval iff it carries a
// "needs_approval" metadata key.
func approvalNeeds(e Event) bool {
	_, ok := e.Metadata["needs_approval"]
	return ok
}

// approvalEvent is the shared fixture: a balance question tagged for approval on c1.
func approvalEvent() Event {
	return Event{
		ConversationID: "c1",
		UserID:         "alice",
		Text:           "what is my balance?",
		Metadata:       map[string]string{"needs_approval": "true"},
	}
}

// TestGateSuspendsOnApproval: an approval-tagged submit suspends instead of running.
func TestGateSuspendsOnApproval(t *testing.T) {
	susp := NewInMemorySuspensionService()
	gate := NewHumanGate(NewBankingRuntime(), susp, approvalNeeds)

	res := gate.Submit(approvalEvent(), 0)

	if res.Path != "awaiting-approval" {
		t.Fatalf("path = %q, want awaiting-approval", res.Path)
	}
	if res.OK {
		t.Fatal("awaiting-approval result should have OK=false")
	}
	if !susp.IsSuspended("c1") {
		t.Fatal("c1 should be suspended after an approval-tagged submit")
	}
}

// TestResumeApprovedReplaysHeldTurn: approving replays the held turn through the runtime,
// which answers the balance question and clears the suspension.
func TestResumeApprovedReplaysHeldTurn(t *testing.T) {
	susp := NewInMemorySuspensionService()
	gate := NewHumanGate(NewBankingRuntime(), susp, approvalNeeds)
	gate.Submit(approvalEvent(), 0)

	res := gate.Resume("c1", true, 10)

	if res.Path != "payments" {
		t.Fatalf("path = %q, want payments", res.Path)
	}
	if !strings.Contains(res.Reply, "1234.56") {
		t.Fatalf("reply %q missing balance amount", res.Reply)
	}
	if susp.IsSuspended("c1") {
		t.Fatal("c1 should no longer be suspended after resume")
	}
}

// TestResumeDenied: denying reports a denial and clears the suspension without running.
func TestResumeDenied(t *testing.T) {
	susp := NewInMemorySuspensionService()
	gate := NewHumanGate(NewBankingRuntime(), susp, approvalNeeds)
	gate.Submit(approvalEvent(), 0)

	res := gate.Resume("c1", false, 10)

	if res.Path != "denied" {
		t.Fatalf("path = %q, want denied", res.Path)
	}
	if res.OK {
		t.Fatal("denied result should have OK=false")
	}
	if susp.IsSuspended("c1") {
		t.Fatal("c1 should no longer be suspended after a denied resume")
	}
}

// TestNormalSubmitPassesThrough: an event with no approval metadata runs as an ordinary
// turn and never suspends.
func TestNormalSubmitPassesThrough(t *testing.T) {
	susp := NewInMemorySuspensionService()
	gate := NewHumanGate(NewBankingRuntime(), susp, approvalNeeds)

	res := gate.Submit(NewEvent("c1", "alice", "what is my balance?"), 0)

	if res.Path != "payments" {
		t.Fatalf("path = %q, want payments", res.Path)
	}
	if !res.OK {
		t.Fatal("a normal balance turn should have OK=true")
	}
	if susp.IsSuspended("c1") {
		t.Fatal("a normal submit should not suspend the conversation")
	}
}

// TestCheckTimeoutsEscalates: a suspension older than the timeout is escalated and cleared;
// younger ones are left pending.
func TestCheckTimeoutsEscalates(t *testing.T) {
	susp := NewInMemorySuspensionService()
	gate := NewHumanGateWithTimeout(NewBankingRuntime(), susp, approvalNeeds, 1000)
	gate.Submit(approvalEvent(), 0)

	if escalated := gate.CheckTimeouts(500); len(escalated) != 0 {
		t.Fatalf("within-timeout sweep escalated %d, want 0", len(escalated))
	}
	if !susp.IsSuspended("c1") {
		t.Fatal("c1 should still be suspended within the timeout")
	}

	escalated := gate.CheckTimeouts(2000)
	if len(escalated) != 1 {
		t.Fatalf("past-timeout sweep escalated %d, want 1", len(escalated))
	}
	if escalated[0].Path != "escalated" {
		t.Fatalf("path = %q, want escalated", escalated[0].Path)
	}
	if susp.IsSuspended("c1") {
		t.Fatal("c1 should be cleared after escalation")
	}
}

// TestResumeNothingPending: resuming a conversation with no held turn reports it.
func TestResumeNothingPending(t *testing.T) {
	susp := NewInMemorySuspensionService()
	gate := NewHumanGate(NewBankingRuntime(), susp, approvalNeeds)

	res := gate.Resume("c1", true, 0)

	if res.Path != "resume" {
		t.Fatalf("path = %q, want resume", res.Path)
	}
	if res.OK {
		t.Fatal("nothing-pending resume should have OK=false")
	}
}
